# Executing supervised learning pipelines with pseudo data for multiple workflows

## Imports

In [6]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, mean_squared_error, root_mean_squared_error
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier

## 1. Predictive Maintenance (Regression)

In [7]:
np.random.seed(42)
maintenance_data = pd.DataFrame({
    'MACHINE_ID': np.arange(1000),
    'HOURS_USED': np.random.randint(500, 2000, 1000),
    'TEMPERATURE': np.random.randint(60, 100, 1000),
    'FAILURE_TIME': np.random.randint(600, 2200, 1000)
})

X_maint = maintenance_data[['HOURS_USED','TEMPERATURE']]
y_maint = maintenance_data['FAILURE_TIME']
X_train, X_test, y_train, y_test = train_test_split(X_maint, y_maint, test_size=0.2, random_state=42)

reg_model = LinearRegression()
reg_model.fit(X_train, y_train)
y_pred = reg_model.predict(X_test)
reg_rmse = root_mean_squared_error(y_test, y_pred)

In [8]:
reg_rmse

460.87276694269895

## 2. Fraud Detection (Classification)

In [9]:
fraud_data = pd.DataFrame({
    'TRANSACTION_ID': np.arange(1000),
    'AMOUNT': np.random.randint(100, 5000, 1000),
    'LOCATION': np.random.choice(['Manila','Cebu','Davao'], 1000),
    'FRAUD_FLAG': np.random.choice([0,1], 1000, p=[0.9,0.1])
})

X_fraud = fraud_data[['AMOUNT','LOCATION']]
y_fraud = fraud_data['FRAUD_FLAG']

preprocessor = ColumnTransformer([
    ('num', StandardScaler(), ['AMOUNT']),
    ('cat', OneHotEncoder(), ['LOCATION'])
])

fraud_pipeline = Pipeline([
    ('preprocess', preprocessor),
    ('clf', RandomForestClassifier(random_state=42))
])

fraud_pipeline.fit(X_fraud, y_fraud)
y_pred_fraud = fraud_pipeline.predict(X_fraud)
fraud_acc = accuracy_score(y_fraud, y_pred_fraud)
fraud_prec = precision_score(y_fraud, y_pred_fraud)
fraud_rec = recall_score(y_fraud, y_pred_fraud)
fraud_f1 = f1_score(y_fraud, y_pred_fraud)

## 3. Customer Analytics (Neural Networks)

In [10]:
customer_data = pd.DataFrame({
    'CUSTOMER_ID': np.arange(1000),
    'AGE': np.random.randint(18,65,1000),
    'PURCHASE_AMOUNT': np.random.randint(50,2000,1000),
    'RESPONSE_TO_PROMO': np.random.choice([0,1], 1000)
})

X_cust = customer_data[['AGE','PURCHASE_AMOUNT']]
y_cust = customer_data['RESPONSE_TO_PROMO']

scaler = StandardScaler()
X_cust_scaled = scaler.fit_transform(X_cust)

nn_model = MLPClassifier(hidden_layer_sizes=(50,), max_iter=500, random_state=42)
nn_scores = cross_val_score(nn_model, X_cust_scaled, y_cust, cv=5)
nn_mean_acc = np.mean(nn_scores)

## 4. Compliance Automation (Decision Trees)

In [13]:
compliance_data = pd.DataFrame({
    'APPLICATION_ID': np.arange(1000),
    'INCOME': np.random.randint(20000,100000,1000),
    'CREDIT_SCORE': np.random.randint(300,850,1000),
    'APPROVED': np.random.choice([0,1], 1000, p=[0.3,0.7])
})

X_comp = compliance_data[['INCOME','CREDIT_SCORE']]
y_comp = compliance_data['APPROVED']

comp_model = DecisionTreeClassifier(max_depth=5, random_state=42)
comp_model.fit(X_comp, y_comp)
y_pred_comp = comp_model.predict(X_comp)
comp_acc = accuracy_score(y_comp, y_pred_comp)

# Results Summary

In [14]:
results = {
    'Predictive Maintenance RMSE': reg_rmse,
    'Fraud Detection Metrics': {
        'Accuracy': fraud_acc,
        'Precision': fraud_prec,
        'Recall': fraud_rec,
        'F1': fraud_f1
    },
    'Customer Analytics NN Accuracy': nn_mean_acc,
    'Compliance Automation Accuracy': comp_acc
}

print("Supervised Learning Pipeline Results Summary:")
for key, value in results.items():
    print(f"{key}: {value}")

Supervised Learning Pipeline Results Summary:
Predictive Maintenance RMSE: 460.87276694269895
Fraud Detection Metrics: {'Accuracy': 0.998, 'Precision': 1.0, 'Recall': 0.9777777777777777, 'F1': 0.9887640449438202}
Customer Analytics NN Accuracy: 0.487
Compliance Automation Accuracy: 0.736


## 🧩 Pipeline Breakdown with Results

### 1. Predictive Maintenance (Regression)
- Data: Machine hours, temperature, failure time.
- Model: Linear Regression.
- Result: RMSE ≈ 460.9 → prediction error in hours.
- Why Needed: Forecast equipment failure to schedule maintenance.
- ELI5: Like guessing when a toy will break based on how long it’s been played with.

### 2. Fraud Detection (Classification)
- Data: Transaction amount, location, fraud flag.
- Model: Random Forest Classifier.
- Results:
    - Accuracy: 99.8%
    - Precision: 1.0 (no false positives)
    - Recall: 97.8% (caught most fraud cases)
    - F1: 0.99
- Why Needed: Detect suspicious transactions in real-time.
- ELI5: Like spotting fake coins in a pile — almost perfect detection.

### 3. Customer Analytics (Neural Networks)
- Data: Age, purchase amount, response to promotions.
- Model: MLP Neural Network.
- Result: Accuracy ≈ 48.7% (weak performance due to random pseudo data).
- Why Needed: Predict customer response to marketing campaigns.
- ELI5: Like guessing if a friend will like a gift — harder without good clues.

### 4. Compliance Automation (Decision Trees)
- Data: Income, credit score, approval decision.
- Model: Decision Tree Classifier.
- Result: Accuracy ≈ 73.6%.
- Why Needed: Automate loan approvals with rule-based logic.
- ELI5: Like following a flowchart to decide if someone can borrow money.

## 🔧 Why Each Pipeline Stage Matters
- Data Ingestion: Collects raw enterprise data (ERP, CRM, IoT).
- Preprocessing: Cleans and standardizes inputs for fairness.
- Model Training: Learns patterns from labeled data.
- Parameter Optimization: Tunes hyperparameters for best performance.
- Evaluation & Validation: Confirms model generalizes beyond training data.
- Monitoring & Maintenance: Detects drift and retrains models for relevance.

## 📌 Critical Lens
- Predictive Maintenance: Regression is interpretable but may miss nonlinear effects → consider Gradient Boosting.
- Fraud Detection: Ensemble methods excel; interpretability tools (SHAP, LIME) needed for compliance.
- Customer Analytics: Neural networks underperform with weak features → richer behavioral data required.
- Compliance Automation: Decision trees are transparent but shallow; ensembles improve robustness.
